tests/player.c

```diff
@@ -536,6 +439,335 @@ START_TEST (test_play_audio_eos)
+  fail_unless (uri != NULL);
+  gst_player_set_uri (player, uri);
+  g_free (uri);
+
+  gst_player_play (player);
+  g_main_loop_run (state.loop);
+
+  g_object_unref (player);
+  g_main_loop_unref (state.loop);
+}
+
+END_TEST;
+
+static void
+switch_streams (GstPlayer * player, GType type)
+{
+  GList *list = NULL;
+  gint current_index = -1;
+  GstPlayerMediaInfo *media_info;
+
+  media_info = gst_player_get_media_info (player);
+  fail_unless (media_info != NULL);
+
+  if (type == GST_TYPE_PLAYER_SUBTITLE_INFO) {
+    GstPlayerSubtitleInfo *sub_info;
+
+    sub_info = gst_player_get_current_subtitle_track (player);
+    fail_unless (sub_info != NULL);
+    current_index = gst_player_stream_info_get_index
+                    ((GstPlayerStreamInfo *) sub_info);
+    list = gst_player_get_subtitle_streams (media_info);
+  } else if (type == GST_TYPE_PLAYER_AUDIO_INFO) {
+    GstPlayerAudioInfo *audio_info;
+
+    audio_info = gst_player_get_current_audio_track (player);
+    fail_unless (audio_info != NULL);
+    current_index = gst_player_stream_info_get_index
+                    ((GstPlayerStreamInfo *) audio_info);
+    list = gst_player_get_audio_streams (media_info);
+  }
+
+  for (; list != NULL; list = list->next) {
+    gint index;
+    GstPlayerStreamInfo *stream = (GstPlayerStreamInfo *) list->data;
+
+    index = gst_player_stream_info_get_index (stream);
+    if (index == current_index)
+      continue;
+
+    {
+      GstPlayerStreamInfo *new_stream = NULL;
+
+      /* set the new track and compare it with index returned from the player */
+      if (type == GST_TYPE_PLAYER_SUBTITLE_INFO) {
+        fail_unless (gst_player_set_subtitle_track (player, index) == TRUE);
+        new_stream = (GstPlayerStreamInfo *)
+                          gst_player_get_current_subtitle_track (player);
+      } else if (type == GST_TYPE_PLAYER_AUDIO_INFO) {
+        fail_unless (gst_player_set_audio_track (player, index) == TRUE);
+        new_stream = (GstPlayerStreamInfo *)
+                          gst_player_get_current_audio_track (player);
+      }
+
+      fail_unless_equals_int (gst_player_stream_info_get_index (new_stream),
+                              index);
+      /* wait for 2 seconds before switching to next stream */
+      sleep (2);
+    }
+  }
+}
+
+static void
+test_play_stream_selection_cb (GstPlayer * player, TestPlayerStateChange change,
+    TestPlayerState * old_state, TestPlayerState * new_state)
+{
+  gint completed = GPOINTER_TO_INT (new_state->test_data);
+
+  /* If pipeline is playing and stream switching is not tested
+   * then switch to next available streams */
+  if (new_state->state == GST_PLAYER_STATE_PLAYING && !completed) {
+    switch_streams (player, GST_TYPE_PLAYER_SUBTITLE_INFO);
+    switch_streams (player, GST_TYPE_PLAYER_AUDIO_INFO);
+    new_state->test_data = GINT_TO_POINTER (completed + 1);
+  }
+  else if (change == STATE_CHANGE_END_OF_STREAM)
```

Comment:
EOS here is also one minute away, that's a bit long if not really needed. Stop after all stream switches were successful